### PR TITLE
pod: frame filters and alloc improvement

### PR DESCRIPTION
**Pull request by ozwaldorf**
## Why

resolves todo for implementing frame filters, and improves some allocations for the inner buffer.

## What

- add frame filter impl
- bring back read_buffer and reserve the len in the inner buffer
- whenever reading a frame, reserve the len again to make sure self.buffer always has at least 148 bytes available (max frame size)
- also added some minor cleanup in example server and ufdp-bench-parser
- improved code comments and added some todos

## Checklist

- [-] I have made corresponding changes to the tests
- [-] I have made corresponding changes to the documentation
- [x] I have run the app using my feature and ensured that no functionality is broken

### CODE REVIEW

**Review by kckeiks — COMMENTED** [review]
LGTM.

Nit:
`// Doc and comment consistency, preferably capitalize the first word and end should be marked with a period.`

File: crates/ursa-pod/src/server.rs [hunk 1]
```diff
@@ -180,26 +176,21 @@ impl<S: AsyncWrite + AsyncRead + Unpin, B: Backend> UfdpHandler<S, B> {
...
                 }
-                Some(f) => {
-                    error!("Terminating asdf, unexpected frame: {f:?}");
-                    self.conn
-                        .termination_signal(Some(Reason::UnexpectedFrame))
-                        .await?;
-                    return Err(UrsaCodecError::UnexpectedFrame(f.tag().unwrap()));
-                }
                 None => return Err(UrsaCodecError::Unknown),
+                Some(_) => unreachable!(),
```
Comment (qti3e) [inline crates/ursa-pod/src/server.rs:189]: Don't trust the client to oblige to the protocol. we should never panic here.

**Review by qti3e — CHANGES_REQUESTED** [review]

File: crates/ursa-pod/src/server.rs (hunk 1 above)
Comment (ozwaldorf) [reply crates/ursa-pod/src/server.rs:189]: the frame filter should guarantee unreachable there

**Review by qti3e — APPROVED** [review]
eyval.

**Pull request merged**